File: src/haven/purus/pbot/api/PBotGobAPI.java
package haven.purus.pbot.api;

import haven.*;

import java.util.*;
import java.util.regex.Pattern;

public class PBotGobAPI {
	PBotSession pBotSession;

	PBotGobAPI(PBotSession pBotSession) {
		this.pBotSession = pBotSession;
	}

	/**
	 * Get gobs matching the specific regex pattern
	 * @param resname regex pattern
	 * @return List of PBotGobs with resname matching the given regex pattern
	 */
	public List<PBotGob> getGobsByResname(String resname) {
		List<PBotGob> ret = new ArrayList<>();
		Pattern pat = Pattern.compile(resname);
		List<Gob> gobs = new ArrayList<>();
		synchronized(pBotSession.gui.ui.sess.glob.oc) {
			pBotSession.gui.ui.sess.glob.oc.forEach(gobs::add);
		}
		for(Gob gob : gobs) {
			Resource res;
			while(true) {
				try {
					res = gob.getres();
					break;
				} catch(Loading l) {
					PBotUtils.sleep(10);
				}
			}
			if(res != null && pat.matcher(res.name).matches())
				ret.add(new PBotGob(gob, pBotSession));
		}
		return ret;
	}

	/**
	 * Get the gob player if it exists
	 * @return Player gob or null if not found
	 */
	public PBotGob getPlayer() {
		Gob gob = pBotSession.gui.map.player();
		if(gob == null)
			return null;
		else
			return new PBotGob(gob, pBotSession);
	}

	/**
	 * Get closest gob matching the specific regex pattern
	 * @param resname regex pattern
	 * @return Closest PBotGob from player with resname matching the given regex pattern
	 */
	public PBotGob getClosestGobByResname(String resname) {
		double bestDistance = Double.MAX_VALUE;
		PBotGob bestGob = null;
		Pattern pat = Pattern.compile(resname);
		List<Gob> gobs = new ArrayList<>();
		synchronized(pBotSession.gui.ui.sess.glob.oc) {
			pBotSession.gui.ui.sess.glob.oc.forEach(gobs::add);
		}
		for(Gob gob : gobs) {
			if(gob.isPlayer())
				continue;
			Resource res;
			while(true) {
				try {
					res = gob.getres();
					break;
				} catch(Loading l) {
					PBotUtils.sleep(10);
				}
			}
			PBotGob pl = getPlayer();
			if(pl == null)
				continue;
			PBotGob pgob = new PBotGob(gob, pBotSession);
			if(res != null && res.name.startsWith("gfx/kritter/horse/") && pgob.dist(pl) == 0)
				continue;
			if(res != null && pat.matcher(res.name).matches() && pgob.dist(pl) < bestDistance) {
				bestDistance = pgob.dist(pl);
				bestGob = pgob;
			}
		}
		return bestGob;
	}

	/**
	 * Returns gob with exactly the given coords or null if not found
	 * @param x x-coordinate of the gob
	 * @param y y-coordinate of the gob
	 * @return Gob with coordinates or null
	 */
	public PBotGob getGobWithCoords(double x, double y) {
		while(true) {
			try {
				synchronized(pBotSession.gui.ui.sess.glob.oc) {
					for(Gob gob : pBotSession.gui.ui.sess.glob.oc) {
						if(Math.abs(gob.getc().x - x) < 0.01 && Math.abs(gob.getc().y - y) < 0.01)
							return new PBotGob(gob, pBotSession);
					}
				}
			} catch(Loading l) {
				try {
					l.waitfor();
				} catch(InterruptedException e) {
					e.printStackTrace();
				}
				continue;
			}
			return null;
		}
	}

	/**
	 * Find object by ID
	 * @param id ID of the object to look for
	 * @return Object, or null if not found
	 */
	public PBotGob findGobById(long id) {
		Gob gob = pBotSession.gui.ui.sess.glob.oc.getgob(id);
		if(gob == null)
			return null;
		else
			return new PBotGob(pBotSession.gui.ui.sess.glob.oc.getgob(id), pBotSession);
	}

	/**
	 * Used to register gob callbacks see examples/gobname.py for usage
	 * @param cb callback
	 */
	public void selectGob(Callback cb) {
		pBotSession.PBotUtils().sysMsg("To select gob, alt + left click");
		synchronized(pBotSession.gui.map.gobCbQueue) {
			pBotSession.gui.map.gobCbQueue.add(new Pair<>(cb, pBotSession));
		}
	}

	// Sorts coordinate array to efficient zig-zag-like sequence for farming etc.
	private static class CoordSort implements Comparator<PBotGob> {
		public int compare(PBotGob a, PBotGob b) {
			if (a.getCoords().floor().x == b.getCoords().floor().x) {
				if (a.getCoords().floor().x % 2 == 0)
					return Integer.compare(b.getCoords().floor().y, a.getCoords().floor().y);
				else
					return Integer.compare(a.getCoords().floor().y, b.getCoords().floor().y);
			} else
				return Integer.compare(a.getCoords().floor().x, b.getCoords().floor().x);
		}
	}


	/**
	 * Returns a list of gobs in the rectangle between A and B points
	 * @param ax x-coord of A point
	 * @param ay y-coord of A point
	 * @param bx x-coord of B point
	 * @param by y-coord of B point
	 * @return List of gobs in the area, sorted to zig-zag pattern
	 */
	public ArrayList<PBotGob> gobsInArea(double ax, double ay, double bx, double by) {
		// Initializes list of crops to harvest between the selected coordinates
		ArrayList<PBotGob> gobs = new ArrayList<PBotGob>();
		double bigX = Math.max(ax, bx);
		double smallX = Math.min(ax, bx);
		double bigY = Math.max(ay, by);
		double smallY = Math.min(ay, by);
		synchronized(pBotSession.gui.ui.sess.glob.oc) {
			for(Gob gob : pBotSession.gui.ui.sess.glob.oc) {
				if(gob.rc.x <= bigX && gob.rc.x >= smallX && gob.getres() != null && gob.rc.y <= bigY
						&& gob.rc.y >= smallY) {
					gobs.add(new PBotGob(gob, pBotSession));
				}
			}
		}
		gobs.sort(new CoordSort());
		return gobs;
	}

}

File: scripts/__pbot/PBotGobAPI.py
from typing import Optional, Callable, List

from .PBotGob import PBotGob


class PBotGobAPI(object):
    def __init__(self, session):
        self.__session = session

    ## Get gobs matching the specific regex pattern
    # @param resname regex pattern
    # @return List of PBotGobs with resname matching the given regex pattern
    def get_gobs_by_resname(self, resname: str) -> List[PBotGob]:
        return [PBotGob(x) for x in self.__session.PBotGobAPI().getGobsByResname(resname)]

    ## Get the player gob if it exists
    # @return player gob ot None if not found
    def get_player(self) -> PBotGob:
        gob = self.__session.PBotGobAPI().getPlayer()
        return PBotGob(gob) if gob is not None else None


    ## Get closest gob matching the specific regex pattern
    # @param resname regex pattern
    # @return closest PBotGob
    def get_closest_gob_by_resname(self, resname: str) -> Optional[PBotGob]:
        gob = self.__session.PBotGobAPI().getClosestGobByResname(resname)
        return PBotGob(gob) if gob is not None else None

    ## Returns gob with exactly the given coords or None if not found
    # @param x x-coordinate of the gob
    # @param y y-coordinate of the gob
    # @return gob with coordinates given or null
    def get_gob_by_coords(self, x: float, y: float) -> Optional[PBotGob]:
        gob = self.__session.PBotGobAPI().getGobWithCoords(x, y)
        return PBotGob(gob) if gob is not None else None

    ## Find object by id
    # @param id id of the object to look for
    # @return object with id or None if not found
    def find_gob_by_id(self, id: int) -> Optional[PBotGob]:
        gob = self.__session.PBotGobAPI().findGobById(id)
        return PBotGob(gob) if gob is not None else None

    ## Used to register gob callbacks
    # @param cb callback called with the gob
    def select_gob(self, cb: Callable[[PBotGob], any]):
        self.__session.PBotGobAPI().selectGob(_SelectGobCb(cb))

    ## Returns a list of gobs in the rectangle between A and B points
    # @param ax x-coord of A point
    # @param ay y-coord of A point
    # @param bx x-coord of B point
    # @param by y-coord of B point
    # @return list of PBotGobs in the area
    def gobs_in_area(self, ax: float, ay: float, bx: float, by: float) -> List[PBotGob]:
        return [PBotGob(x) for x in self.__session.PBotGobAPI().gobsInArea(ax, ay, bx, by)]


class _SelectGobCb(object):
    def __init__(self, cb: Callable[[PBotGob], any]):
        self.cb = cb

    def callback(self, gob):
        self.cb(PBotGob(gob))

    class Java:
        implements = ["haven.purus.pbot.api.Callback"]